libyul/AST.h

```diff
@@ -69,14 +73,20 @@ class LiteralValue {
 struct Literal { langutil::DebugData::ConstPtr debugData; LiteralKind kind; LiteralValue value; };
 /// External / internal identifier or label reference
 struct Identifier { langutil::DebugData::ConstPtr debugData; YulName name; };
+/// Builtin function
+struct Builtin { langutil::DebugData::ConstPtr debugData; BuiltinHandle handle; };
```

Comment:
```suggestion
struct BuiltinName { langutil::DebugData::ConstPtr debugData; BuiltinHandle handle; };
```

Reply:
On a related note, I was confused between `FunctionName` and `FunctionNameIdentifier` at first, the current naming is not very intuitive to me. I think that `FunctionNameIdentifier` might be better off named `FunctionHandle`. 

I'd also put it in `ASTForward.h`. Or maybe even `Builtins.h`? Having it here made me look for a connection with other AST nodes and apparently there isn't any.

Reply:
I wish all these AST structs had a `Node` suffix. Our current naming convention for them is just too generic and makes naming things in a non-confusing way too hard.

Reply:
> On a related note, I was confused between `FunctionName` and `FunctionNameIdentifier` at first, the current naming is not very intuitive to me. I think that `FunctionNameIdentifier` might be better off named `FunctionHandle`.

Yeah that works! `FunctionHandle` as union over `YulName`, `BuiltinHandle`, and `VerbatimHandle`. Could in principle even typedef `UserDefinedFunctionHandle` as `YulName`. 

> I'd also put it in `ASTForward.h`. Or maybe even `Builtins.h`? Having it here made me look for a connection with other AST nodes and apparently there isn't any. They are not using it at all and it's not using them.

It is a lightweight version of the `FunctionName` variant to be used as key type in maps and such. That is also why I called it `FunctionNameIdentifier` in the first place. It does exceed the scope of builtins, so I'd be more inclined to put it into `ASTForward.h`. 

> I wish all these AST structs had a Node suffix. Or were in a node namespace or something. Our current naming convention for them is just too generic and makes naming things in a non-confusing way harder than it needs to be.

True! We can do that in a follow up, I think the PR is large enough already:)

Reply:
> True! We can do that in a follow up, I think the PR is large enough already:)

Yeah, definitely don't do this in this PR. I'm not sure if we even should do this at all. This seems like the source of our naming problems here, but it was just a quick idea, not something that I though through and it would be quite invasive (lots of conflicts with existing PRs), so I wouldn't want to have to revert that :)